src/providers/Y2KEarthquakeCarouselInsuranceProvider.sol

```diff
@@ -192,33 +223,49 @@ contract Y2KEarthquakeCarouselInsuranceProvider is IInsuranceProvider {
         uint256 amountPremium,
         uint256 amountCollateral
     ) external {
-        (,,address underlyingAsset) = carouselFactory.marketIdInfo(marketId);
+        (, , address underlyingAsset) = carouselFactory.marketIdInfo(marketId);
         address[2] memory vaults = carouselFactory.getVaults(marketId);
-        IERC20(underlyingAsset).safeTransferFrom(msg.sender, address(this), amountPremium + amountCollateral);
-        IERC20(underlyingAsset).safeApprove(vaults[0], amountPremium);
-        IERC20(underlyingAsset).safeApprove(vaults[1], amountCollateral);
+        IERC20(underlyingAsset).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amountPremium + amountCollateral
+        );
 
         uint256 nextEpochId = nextEpoch(ICarouselExtended(vaults[0]));
-        ICarouselExtended(vaults[0]).deposit(nextEpochId, amountPremium, msg.sender);
-        ICarouselExtended(vaults[1]).deposit(nextEpochId, amountCollateral, msg.sender);
+        if (amountPremium > 0) {
+            IERC20(underlyingAsset).safeApprove(vaults[0], amountPremium);
+            ICarouselExtended(vaults[0]).deposit(
+                nextEpochId,
+                amountPremium,
+                msg.sender
+            );
+        }
+        if (amountCollateral > 0) {
+            IERC20(underlyingAsset).safeApprove(vaults[1], amountCollateral);
+            ICarouselExtended(vaults[1]).deposit(
+                nextEpochId,
+                amountCollateral,
+                msg.sender
+            );
+        }
     }
 
     /**
      * @notice Claims payout for the resolved epochs.
      * @param marketId Market Id
      */
+    // NOTE: The nextEpochIndexToClaim is called by the sender and this changes nextEpochIndexToClaim
+    // NOTE: Could this in theory malicious push the epoch index forward?
```

Comment:
This is a good catch! I missed this point when we were turning insurance provider into unpermissioned generalized contract.